src/Core/Geometry/IndexedGeometry.cpp

```diff
@@ -0,0 +1,254 @@
+}
+
+//////////////////////////////////////////////////////////////////////
+//////////////////////////////////////////////////////////////////////
+
+GeometryIndexLayerBase&
+MultiIndexedGeometry::getFirstLayerOccurrenceWithLock( const LayerSemantic& semanticName ) {
+    for ( auto& [key, value] : m_indices )
+    {
+        if ( key.first.find( semanticName ) != key.first.end() )
+        {
+            CORE_ASSERT( !value.first, "try to get already locked layer" );
+            value.first = true;
+            return *( value.second );
+        }
+    }
+    throw std::out_of_range( "Layer entry not found" );
+}
+
+GeometryIndexLayerBase&
+MultiIndexedGeometry::getFirstLayerOccurrenceWithLock( const LayerSemanticCollection& semantics ) {
+    for ( auto& [key, value] : m_indices )
+    {
+        if ( key.first == semantics )
+        {
+            CORE_ASSERT( !value.first, "try to get already locked layer" );
+            value.first = true;
+            return *( value.second );
+        }
+    }
+    throw std::out_of_range( "Layer entry not found" );
+}
+
+GeometryIndexLayerBase& MultiIndexedGeometry::getLayerWithLock( const LayerKeyType& layerKey ) {
+    auto& p = m_indices.at( layerKey );
+    CORE_ASSERT( !p.first, "try to get already locked layer" );
+    p.first = true;
+    return *( p.second );
+}
+
+//////////////////////////////////////////////////////////////////////
+//////////////////////////////////////////////////////////////////////
+
+void MultiIndexedGeometry::unlockFirstLayerOccurrence( const LayerSemantic& semanticName ) {
+    for ( auto& [key, value] : m_indices )
+    {
+        if ( key.first.find( semanticName ) != key.first.end() )
+        {
+            CORE_ASSERT( value.first, "try to release unlocked layer" );
+            value.first = false;
+            notify();
+            return;
+        }
+    }
+    throw std::out_of_range( "Layer entry not found" );
+}
+
+void MultiIndexedGeometry::unlockFirstLayerOccurrence( const LayerSemanticCollection& semantics ) {
+    for ( auto& [key, value] : m_indices )
+    {
+        if ( key.first == semantics )
+        {
+            CORE_ASSERT( value.first, "try to release unlocked layer" );
+            value.first = false;
+            notify();
+            return;
+        }
+    }
+    throw std::out_of_range( "Layer entry not found" );
+}
+
+void MultiIndexedGeometry::unlockLayer( const LayerKeyType& layerKey ) {
+    auto& p = m_indices.at( layerKey );
+    CORE_ASSERT( p.first, "try to release unlocked layer" );
+    p.first = true;
+    notify();
+}
+
+//////////////////////////////////////////////////////////////////////
+//////////////////////////////////////////////////////////////////////
+
+bool MultiIndexedGeometry::addLayer( std::unique_ptr<GeometryIndexLayerBase>&& layer,
+                                     const std::string& layerName ) {
+    LayerKeyType key = std::make_pair( layer->semantics(), layerName );
```

Comment:
(not tested, if ok need to be done for most of the make_pair)

Since there is a using def for the pair, why not use directly
```suggestion
    LayerKeyType key { layer->semantics(), layerName };
```